### PR TITLE
Added text logging features

### PR DESCRIPTION
**Pull request by ryanelandt**
This pull request expands Drake's text logging tools to handle multiple loggers whose names are not known a priori. Named loggers can be created at the command line or in code. The logging level (debug, trace, etc) of loggers are set independently. A logger's level can change in a function if more data is desired. Attempting to log to a non-existant logger doesn't error and logs to a default logger instead. All data is logged to the same sink. Messages from different loggers should be stagger by line and not commingle. The goal of this PR is to give users the tools to discover the most valuable ways to log their data.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955)

### CODE REVIEW

**Comment by sherm1** [comment]
@drake-jenkins-bot add to whitelist

**Comment by edrumwri** [comment]
+@edrumwri

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955#-:-LEUrijj59-xbpm9wIUI:beamsal)*

File: common/text_logging.cc [hunk 1]
```diff
@@ -6,31 +6,96 @@
...
+  if (
+        !want_unchanged &&
+        !want_trace &&
+        !want_debug &&
+        !want_info &&
+        !want_warn &&
+        !want_err &&
+        !want_critical &&
+        !want_off
+      ) {
```
Comment (sprax) [inline common/text_logging.cc]: so glad to see these "missing" levels added!  Thanks.

**Comment by ryanelandt** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[common/text_logging.cc, line 28 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LES4MN7IMJW9tGlQZNT-r1-28:-LEW4LwFBmywktHDsr1-:b-bavq6s) ([raw file](https://github.com/robotlocomotion/drake/blob/0748b710fb5230f16cc047ae800dd898a7ea5a7a/common/text_logging.cc#L28)):*
<details><summary><i>Previously, sprax (Sprax Lines) wrote…</i></summary><blockquote>

so glad to see these "missing" levels added!  Thanks.
</blockquote></details>

You're welcome.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955)*

**Comment by edrumwri** [comment]
Ryan, your logger needs some tests- you can update the existing ones in common/tests/text_logging_test.cc

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[common/text_logging.h, line 11 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWmaw83j1uZlfVil9s:-LEWmaw83j1uZlfVil9t:b-2igmpt) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.h#L11)):*
> ```objc
> from commandline arguments with drake::HandleSpdlogGflags. One can log
> to loggers whose existence is unknown at compile time using
> drake::ReturnLoggerPreferablyNamed. Once you've included this file, the
> ```

Please document why ...PreferablyNamed and ...AlwaysNamed are provided (distinguish between them).

---

*[common/text_logging.h, line 146 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWmQo9-aXFmjrKHpAN:-LEWmQo9-aXFmjrKHpAO:blm58et) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.h#L146)):*
> ```objc
> /// Retrieve an instance of a logger to use for logging; for example:
> ///   `drake::log()->info("potato!")`
> logging::logger* log();
> ```

Why not add a function "log(const std::string&)" which would log to the particular named logger?

---

*[common/text_logging.h, line 156 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWk08M-U6IOCyms_X8:-LEWk08NAL5O7lNz6hD_:b-tmaea6) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.h#L156)):*
> ```objc
> ///  drake::SetLoggerLevel(logger_console, "debug");
> /// </pre>
> void SetLoggerLevel(const std::shared_ptr<logging::logger> &logger_ptr,
> ```

Nit: ampersand goes by the type, not the variable name

---

*[common/text_logging.cc, line 11 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWkQ957NCBmUGJFNA-:-LEWkQ967XuR-mN-YMiY:b-wfburz) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.cc#L11)):*
> ```cpp
> void SetLoggerLevel(const std::shared_ptr<logging::logger>& logger_ptr,
>                     const std::string& logger_level) {
>   const bool want_unchanged = (logger_level == "unchanged");
> ```

Oof! This function is calling for a enum type! You could use the spdlog enum if it's not necessary to be able to change the logging level to unchanged (what functionality does that buy you?)

---

*[common/text_logging.cc, line 109 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWnDil1BIrxi0xlOKg:-LEWnDim4v7RHGkrCO4m:bb3cyi2) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.cc#L109)):*
> ```cpp
> }
> 
> std::shared_ptr<logging::logger> ReturnLoggerPreferablyNamed(
> ```

I'm missing something- why is this (and the functions below) returning a null pointer?

---

*[common/text_logging_gflags.h, line 32 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWnetz8_4BYggAXX5j:-LEWnetz8_4BYggAXX5k:bprqesa) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging_gflags.h#L32)):*
> ```objc
> /// it to trace and creates a logger named "logger_b" and sets it to info.
> /// <pre>
> ///  ./compiled_project --spdlog_level=off logger_a trace logger_b info
> ```

What do you think of separating the named logger and its level with commas?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955#-:-LEWm45tAyJIwMcsp985:b53lvd)*

**Comment by jwnimmer-tri** [comment]
As the area owner for this component, I'll self-assign additional review.

Factoring out a function to do `set_level` by string is fine, and teaching the gflags processing and help string about the new values is fine.  Maybe that could be done first, as its own PR?  That's just simple nuts and bolts.

The new feature of supporting named log channels requires more discussion.  Could you elaborate more about the motivation here, such as whether we expect to use this within Drake, or whether only downstream project would use it?

The text logging in Drake is simple on purpose.  If we want to make it more complicated, we need to develop the user story that motivates the work.

**Comment by ryanelandt** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[common/text_logging.cc, line 109 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWnDil1BIrxi0xlOKg:-LEWoxg5BW7JlfsgMwPF:b-escpc6) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.cc#L109)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I'm missing something- why is this (and the functions below) returning a null pointer?
</blockquote></details>

I stubbed out these function so that Drake code will still work if Spdlog isn't installed. This is similar to what was done in text_logging.h.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955)*

**Comment by edrumwri** [comment]
I can do that since I requested this feature. Logging is easily capable of producing too much or too little information. For example, I may be debugging the LCP solver. Logging for multibody dynamics, collision detection, integration, etc. would prove to be a distraction.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955#-:-LEWr_3v8hwOOV9vTx1R:br1k888)*

**Comment by jwnimmer-tri** [comment]
Alright.  We should video chat then, next week, to choose a path forward (ping a time on my calendar).  The short version is:
- (1) Drake is a library; its logging is simple in part because the more complex that we make that subsystem, the more surface area it is for system integrators to bind our logger calls into their logging own back-end(s).  We cannot assume that everyone who uses Drake is using spdlog as the final destination for their logs.
- (2) Streams of unstructured text output is a relatively low-value way to diagnose behaviors in code.  We should endeavor to build and apply better development / debugging practices, instead of relying on the text output crutch.  Keeping the logging system simple helps us steer people away from it.

**Comment by edrumwri** [comment]
Yes, let's do that. I'm unconvinced by (2) though- I've found logging to be a valuable debugging mechanism complementary to other debugging tools, particularly for numerical algorithms. Other Drake developers feel this way also.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955#-:-LEWtP828wErvqjqJPBz:b2pkpzh)*

**Comment by ryanelandt** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[common/text_logging.h, line 146 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWmQo9-aXFmjrKHpAN:-LEWqeLN7odgCtCcJR8X:b-b06ecf) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.h#L146)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Why not add a function "log(const std::string&)" which would log to the particular named logger?
</blockquote></details>

As it is currently set up, "drake::log()" evaluates once and logs to the same logger every time. This is fine for now because there is only one logger. I tried to make the function you suggested, but it logs to the logger associated with its first input irrespective of what it's called with.

I did not remove "never_destroyed" in order to make a version that looks up loggers by name every time for speed reasons. Also, I think what I did is consistent how the makers of Spdlog intend for people to use it (which for someone with my C++ experience is probably a better than the alternative).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955)*

**Comment by jwnimmer-tri** [comment]
I should also note that even if we end up agreeing that channelized logs are worth it (shooting down (2) and deciding to pay the cost of (1)), then likely we reach:

- (3) if we want channelized logs, we switch from using spdlog to instead use a thorough and well-worn framework for it (one of the log4cpp or log4cpp family, perhaps), rather than putting band-aids on spdlog.

In that sense, the parts of this PR that try to add channelization are probably not worth iterating on with fixups / reviewable.

**Comment by edrumwri** [comment]
I thought spdlog has seemed reasonable thus far, but you clearly have a differing (and more informed) opinion. In any case, much of what Ryan is doing is getting experience with our development process before his internship starts formally. I'm ok with the "regions" (another name for channels) not getting merged in as long as he is learning from the dialogue.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[common/text_logging.cc, line 109 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8955#-LEWnDil1BIrxi0xlOKg:-LEWuaCn4jYwX5Q6h9xt:b-4nqwob) ([raw file](https://github.com/robotlocomotion/drake/blob/4623f04975c2635f01e7480a1655e3b61fee298b/common/text_logging.cc#L109)):*
<details><summary><i>Previously, ryanelandt wrote…</i></summary><blockquote>

I stubbed out these function so that Drake code will still work if Spdlog isn't installed. This is similar to what was done in text_logging.h.
</blockquote></details>

Got it- please document this then.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955#-:-LEWwsdh8VOCOJlQF7ei:b8k2k7j)*

**Comment by jwnimmer-tri** [comment]
For the record, my thoughts from discussions yesterday:
- Teaching the command-line (and/or helper functions) about all of spdlog's level strings is fine, and could go in early as a first PR.
- My request for the downstream user of Drake is that they have a single method that redirects the sink for all of Drake's loggers / regions, whether the loggers have already been created, or ones that might be created later after a simulation starts.  This could be a method on spdlog directly, or some sugar offered by Drake.  We are not sure how difficult this is with spdlog, but we did not immediately know of a built-in way to do this.
- I am going to look at what it would look like to pivot to use log4cpp in Drake.

**Comment by jwnimmer-tri** [comment]
> My request for the downstream user of Drake is that they have a single method that redirects the sink for all of Drake's loggers / regions, ...

In Drake, if we always make our new region loggers pointing to the same `dist_sink` singleton, then the user could call `add_sink` and `remove_sink` on that `dist_sink` to redirect all messages.  By default, we would latch-init our `dist_sink` singleton to delegate to the `stderr_sink` singleton, but users could then easily swap it out.  On a quick code skim at least, this seems to be like a robust way to offer named "regions".

**Comment by jwnimmer-tri** [comment]
FYI Sample code for sink reconfiguration at https://github.com/jwnimmer-tri/drake/commits/spdlog_dist_sink.

**Comment by edrumwri** [comment]
As usual, Jeremy, you rock!

---

Review status: 6 unresolved discussions, LGTM missing from assignee jwnimmer-tri, LGTM missing from assignee edrumwri, platform LGTM missing

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8955#-:-LEtSHKO7gV1ADo6sED4:bvol4f4)*

**Comment by jwnimmer-tri** [comment]
Closing for lack of interest / updates.  If someone wants to re-open this again later once its merged up to master, feel free.

**Pull request closed without merging**